File: 20-ValidParentheses/20-ValidParentheses.py
# Last updated: 09/10/2025, 09:40:27
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        lst = list(s)
        for i in range (len(lst)):
            stack.append(lst[i])
            if len(stack) > 1:
                if (stack[-2] == "(" and stack[-1] == ")"):
                    stack.pop()
                    stack.pop()
                elif (stack[-2] =="[" and stack[-1] =="]"):
                    stack.pop()
                    stack.pop()
                elif (stack[-2] == "{" and stack[-1] =='}'):
                    stack.pop()
                    stack.pop()
        if len(stack) == 0:
            return True
        return False
        
            

File: 20-ValidParentheses/20-ValidParentheses.cpp
// Last updated: 03/03/2026, 20:07:18
class Solution {
public:
    bool isValid(string s) {
        std::stack<char> st;
        for (char c : s) {
            if (c == '(' || c == '[' || c == '{') {
                // 1. Push all opening brackets onto the stack
                st.push(c);
            } else {
                // 2. Encountered a closing bracket
                // Check if stack is empty or the top doesn't match
                if (st.empty() || 
                    (c == ')' && st.top() != '(') ||
                    (c == ']' && st.top() != '[') ||
                    (c == '}' && st.top() != '{')) 
                {
                    return false; // Mismatch or no opening bracket found
                }
                // 3. Match found, remove the opening bracket
                st.pop(); 
            }
        }
        
        // 4. The string is valid only if the stack is completely empty at the end.
        return st.empty();
    }
};